### ISSUE TITLE
Incorect handling of user cancellation 3d secure auth flow

### ISSUE THREAD

**Issue by mzygar** [stripe/stripe-ios#1268]
## Summary
I’m having some issues with the handling the cancel button when paying by card using 3DSecure flow. 
It is correctly handled in the `STPPaymentHandler.confirmPayment`- the action received in the callback is `.cancelled`. However, we initiate the process from `STPPaymentContext.didCreatePaymentResult`, get the `paymentMethod` ’s ID and continue with the intent process. To properly close the loop, I need to call the `completion` .  Then `STPPaymentContext.paymentContext:didFinishWithStatus` is called.

In this scenario, when I call `completion(nil)`  the `status` received  in `paymentContext:didFinishWithStatus` is `.success` , when I call it `completion(error)` the `status` received is `.error`. I think this might be a bug, as I’d expect to receive `.userCancellation`.
Do you think it is the correct behaviour of the SDK?

## Code to reproduce
The simplified flow looks like this:
1. Initial payment using `STPPaymentContext` and confirm paymentIntent
```swift
//in class implementing STPPaymentContextDelegate
func paymentContext(_ paymentContext: STPPaymentContext, didCreatePaymentResult paymentResult: STPPaymentResult, completion: @escaping STPErrorBlock) {
    output.completion = completion
    // ...
    //get Client secret from the server
    // ...
    let params = STPPaymentIntentParams(clientSecret: paymentIntentDetails.paymentIntentClientSecret)
    params.paymentMethodId = paymentResult.paymentMethod.stripeId
    let paymentManager = STPPaymentHandler.shared()
    paymentManager.confirmPayment(params, with: authenticationContext) { action, paymentIntent, error in
        switch action {
        case .canceled:
            self.output?.paymentIntentCancelled() // as expected
        case .failed:
            //...
        case .succeeded:
            //...
    }
}
```

2. Complete handling the result
```swift
func paymentIntentCancelled() {
    completion?(nil) //or completion?(Error())
}
```

3. Get the status in:
```swift
//in class implementing STPPaymentContextDelegate
func paymentContext(_ paymentContext: STPPaymentContext, didFinishWith status: STPPaymentStatus, error: Error?) {
    switch status { }
}
```
* `status` is `.success` if in previous step we called `completion(nil)`
* `status` is `.error` if in previous step we called `completion(Error)`

Expected status: `.userCancellation`

## iOS version
12.4

## Installation method
Cocoapods

## SDK version
16.0.2

## Other information
Happens in both 3DSecure and 3DSecure v2 flows

**Comment by yuki-stripe** [comment]
Hi @mzygar, thanks for reporting.  It sounds like the issue here is that the completion block in `paymentContext:didCreatePaymentResult:completion:` doesn't let you tell it the user canceled (it only accepts success or error).  I think the fix is to change the completion block type to support calling it with `completion(.canceled)` or something like that.  

As a workaround, when the user cancels, you could pass a special error representing "user canceled" to the completion block.  You'll receive the error you passed in when `paymentContext:didFinishWithStatus:error:` is called.

rough example of what I mean:
```
func paymentIntentCancelled() {
    completion?(MyCancelationError())
}

//in class implementing STPPaymentContextDelegate
func paymentContext(_ paymentContext: STPPaymentContext, didFinishWith status: STPPaymentStatus, error: Error?) {
    switch status { 
        case .error:
            if let error = error as? MyCancelationError {
                // This isn't an error, the user actually canceled, do your cancelation logic.
            }
    }
}
```

It's ugly, but at least you'll be able to detect the user cancellation case until we get a proper fix in.

**Comment by yuki-stripe** [comment]
This is fixed in 17.0.0.  Pass `completion(.userCancellation, nil)` in the cancellation case.
